### ISSUE TITLE
HDUList.__contains__ does not work for integer arguments

### ISSUE THREAD

**Issue by mcara** [astropy/astropy#7269]
It has been mentioned that one can test for presence of a specific extension in an `HDUList` by doing something like this: `'SCI' in HDUList`. For example,

```python
>>> hdulist = fits.open('somefile_flt.fits')
>>> 'SCI' in hdulist
True
```

However, if extension is an integer number, this does not work. For example, for this particular file `len(hdulist)` is 7 but:

```python
>>> 2018 in hdulist
True # <-- should be False!!!
```

The reason for this is that `index_of()` used in `__contains__()` (see https://github.com/astropy/astropy/blob/9626265d77b8a21c113615c08bc6782deb52eaed/astropy/io/fits/hdu/hdulist.py#L329) always returns the integer passed as its argument regardless of whether or not that integer extension number exists in the list.

In my opinion the check above (`2018 in hsulist`) should work and return `False` because "integer extensions numbers" are as valid as string extension names (and, optionally, versions) for identifying specific extensions.

While I could fix this, I would like to see what is others opinion about this issue.

**Comment by mcara** [comment]
In addition, even though `HDUList.__contains__()` contains a docstring - see https://github.com/astropy/astropy/blob/9626265d77b8a21c113615c08bc6782deb52eaed/astropy/io/fits/hdu/hdulist.py#L325 - it does not appear in the sphinx-generated documentation.

**Comment by pllim** [comment]
c/c @saimn @MSeifert04 @drdavella

**Comment by MSeifert04** [comment]
Docstrings for magic methods don't show up in the docs. That's intended because users aren't supposed to invoke them "manually".

I haven't had the chance to look at the issue though.

**Comment by mcara** [comment]
@MSeifert04 Wasn't [`:special-members:`](http://www.sphinx-doc.org/en/stable/ext/autodoc.html) designed for this? Some examples: https://stackoverflow.com/questions/15903577/include-specific-special-methods-in-sphinx

The thing here is that `HDU`**`List`** is a **list** behaving like a **dict** and sometimes documenting these specifics helps... For example, for a list, inclusion (or "contains") test such as:

```python
a in [a, b, c] # a, b, c - some objects
```
does not check for content of `name` members, that is, it does not work like this: `a.name in [a.name, b.name, c.name]` yet it is partially what *seems to be intended* for `HDUList`. IMO, this kind of unexpected behavior should be documented.

**Comment by mcara** [comment]
In addition, for a normal Python **list**, something like this should work:
```python
>>> hdulist = fits.open('somefile_flt.fits')
>>> hdu0 = hdulist[0]
>>> hdu1 = hdulist[1]
>>> hdu0 in hdulist
False # WHAT???
```

**Comment by saimn** [comment]
> In my opinion the check above (2018 in hdulist) should work and return False because "integer extensions numbers" are as valid as string extension names (and, optionally, versions) for identifying specific extensions.

I think using `in` with an integer should always fail (return `False`), because `in`  should  work only with extension name and HDU instances. But I'm confused with #7281 since it does not exactly address this point. #7281 focus on `index_of` but I think that the behavior of `index_of` with integers is a side effect of being used by `_positive_index_of` which itself is used by get/setitem. So I think that `index_of` should not handle integers, this case should be handled in the calling methods. Very quickly (and not tested), I would do something like this: https://github.com/saimn/astropy/commit/03a861d98a09a9a06fd2ea82089f97fd0131e8e0 (which fixes the `int in hdulist` case).

**Comment by dhomeier** [comment]
Don't know if this is unrelated, coincidental or intentional, but I've also noticed that advanced indexing like `hdulist[[1,2,5,4,-1]]` does not work.

**Comment by pllim** [comment]
> `hdulist[[1,2,5,4,-1]]` does not work.

I don't think it is ever supported (at least not in recent releases)? From the code, if you pass in `slice`, it would grab multiple HDUs. But if it is not `slice`, the code seems to assume you only want a single HDU, as I don't see any code to process list of indices.

https://github.com/astropy/astropy/blob/8686f3eeb1728af907ac2acfac2c7c682b23f597/astropy/io/fits/hdu/hdulist.py#L298-L300

**Comment by mcara** [comment]
As an alternative,
```python
fits.HDUList([hdulist[ext] for ext in [1,2,5,4,-1]])
```

**Comment by dhomeier** [comment]
OK, was just wondering if it there was a reason that this did not work like in a plain `list`. List comprehension is a good workaround, where needed.

**Comment by saimn** [comment]
Python's `list` does not support advanced indexing with a list of indices ;).

**Comment by dhomeier** [comment]
Oh right, that's a Numpy feature. Makes no sense to expect it from any `list`-derived class then.
